server/server-core/src/main/java/software/amazon/smithy/java/server/core/CorsHeaders.java

```diff
@@ -0,0 +1,80 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package software.amazon.smithy.java.server.core;
+
+import static software.amazon.smithy.java.core.schema.TraitKey.CORS_TRAIT;
+
+import java.util.*;
+import software.amazon.smithy.java.logging.InternalLogger;
+import software.amazon.smithy.model.traits.CorsTrait;
+
+public final class CorsHeaders {
+    private static final InternalLogger LOGGER = InternalLogger.getLogger(CorsHeaders.class);
+
+    private CorsHeaders() {} // Prevent instantiation
+
+    public static Map<String, List<String>> of(HttpJob job) {
+        if (!shouldAddCorsHeaders(job)) {
+            return Map.of();
+        }
+
+        String requestOrigin = job.request().headers().firstValue("origin");
+        Optional<String> configuredOrigin = getConfiguredOrigin(job);
+
+        if (!isOriginAllowed(configuredOrigin.orElse(null), requestOrigin)) {
+            return Map.of();
+        }
+
+        Map<String, List<String>> headers = new HashMap<>();
```

Comment:
We're allocating and populating a HashMap only to immediately add its contents to another map. We can avoid the unnecessary allocation by mutating the response headers directly in this method.